modules/core/shared/src/main/scala/io/circe/Cursor.scala

```diff
@@ -0,0 +1,1351 @@
+   */
+  final def succeeded: Boolean =
+    focus.nonEmpty
+
+  /**
+   * Whether or not this cursor has failed.
+   *
+   * A failed cursor will always have an empty focus.
+   */
+  final def failed: Boolean = succeeded === false
+
+  /**
+   * Move from the current position to the root of the JSON structure.
+   *
+   * If the cursor is a [[Cursor#SuccessCursor]], this is the same as calling
+   * [[#up]] until [[#isRoot]] returns true. If the cursor is a has failed,
+   * then this will not move the cursor. The only way to exit a cursor in a
+   * failed state is via the [[#lastCursor]].
+   */
+  @tailrec
+  final def root: Cursor =
+    this match {
+      case cursor if cursor.isRoot =>
+        this
+      case _: FailedCursor =>
+        this
+      case otherwise =>
+        otherwise.up.root
+    }
+
+  /**
+   * Replace the focus.
+   *
+   * @group Modification
+   */
+  final def set(j: Json): Cursor = mapFocus(_ => j)
+
+  /**
+   * If the focus is a JSON array, return its elements.
+   *
+   * @group ArrayAccess
+   */
+  final def values: Option[Iterable[Json]] =
+    focus.flatMap {
+      case Json.JArray(vs) => Some(vs)
+      case _               => None
+    }
+
+  /**
+   * If the focus is a JSON object, return its field names in their original order.
+   *
+   * @group ObjectAccess
+   */
+  final def keys: Option[Iterable[String]] =
+    focus.flatMap {
+      case Json.JObject(o) => Some(o.keys)
+      case _               => None
+    }
+
+  /**
+   * Return to the root of the document.
+   *
+   * @group Access
+   */
+  final def top: Option[Json] =
+    root.focus
+
+  /**
+   * Replay an operation against this cursor.
+   *
+   * @group Utilities
+   */
+  final def replayOne(op: CursorOp): Cursor = op match {
+    case CursorOp.MoveLeft       => left
+    case CursorOp.MoveRight      => right
+    case CursorOp.MoveUp         => up
+    case CursorOp.Field(k)       => field(k)
+    case CursorOp.DownField(k)   => downField(k)
+    case CursorOp.DownArray      => downArray
+    case CursorOp.DownN(n)       => downN(n)
+    case CursorOp.DeleteGoParent => delete
+    case CursorOp.Replace(json)  => set(json)
+  }
+
+  /**
+   * Replay history against this cursor.
+   *
+   * @group Utilities
+   */
+  final def replay(history: History): Cursor =
+    replayCursorOps(history.asList)
+
+  /**
+   * Replay history against the cursor for any `Foldable` of [[CursorOp]].
+   *
+   * @note This method is error prone as it is not clear if the first element
+   *       is the first operation or the last operation. Historically the
+   *       first element was the last operation, because of how Circe
+   *       constructed the history, but it can make intuitive sense to have
+   *       the first element be the first operation. For this reason, this
+   *       method is private.
+   */
+  private[circe] final def replayCursorOps[F[_]: Foldable](history: F[CursorOp]): Cursor =
```

Comment:
I think its probably fine to make this public IF we add a comment explaining that the first element is assumed to be the start of the history and not the end.  I can imagine people reimplementing this themselves because they stored (for whatever reason) a History or list of cursor ops and wanted to apply it to another cursor.  Either open this, or maybe make it possible to construct a history from a list of cursor ops?  IDK, i don't have any strong feelings here... Thoughts?

Reply:
I can make it public and update the comment. I think having `History` as an always correct type, but permitting reply by an arbitrary `Foldable` makes sense.

Reply:
I'd like to make the construction of `History` public, but I don't see any easy way to communicate at the types the opposite behavior of the list backing the type.